pkg/orm/client.go

```diff
@@ -0,0 +1,610 @@
+package orm
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	dmysql "github.com/go-sql-driver/mysql"
+	libModel "github.com/hanfei1991/microcosm/lib/model"
+	cerrors "github.com/hanfei1991/microcosm/pkg/errors"
+	"github.com/hanfei1991/microcosm/pkg/externalresource/resourcemeta"
+	"github.com/hanfei1991/microcosm/pkg/meta/metaclient"
+	"github.com/hanfei1991/microcosm/pkg/orm/model"
+	"github.com/hanfei1991/microcosm/pkg/tenant"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+	"gorm.io/driver/mysql"
+	"gorm.io/gorm"
+)
+
+// TODO: retry and idempotent??
+// TODO: context control??
+
+type TimeRange struct {
+	start time.Time
+	end   time.Time
+}
+
+type Client interface {
+	metaclient.Client
+
+	// Initialize will create all tables for backend operation
+	Initialize(ctx context.Context) error
+
+	// project
+	AddProject(ctx context.Context, project *model.ProjectInfo) error
+	DeleteProject(ctx context.Context, projectID string) error
+	QueryProjects(ctx context.Context) ([]*model.ProjectInfo, error)
+	GetProjectByID(ctx context.Context, projectID string) (*model.ProjectInfo, error)
+
+	// project operation
+	AddProjectOperation(ctx context.Context, op *model.ProjectOperation) error
+	QueryProjectOperations(ctx context.Context, projectID string) ([]*model.ProjectOperation, error)
+	QueryProjectOperationsByTimeRange(ctx context.Context, projectID string, tr TimeRange) ([]*model.ProjectOperation, error)
+
+	// job info
+	UpsertJob(ctx context.Context, job *libModel.MasterMetaKVData) error
```

Comment:
With an RDBMS as a storage, do we still need the upsert method?

Reply:
It's for the easy use of framework metaclient replacement.

When we insert the job info/worker status/resource meta after normal startup or failover, we need to check the existence of the item first, then do insert or update if we don't have the `Upsert` interface.

So here I introduce `Upsert`. And I will refine the implement using `replace into` later.